### ISSUE TITLE
Continuous integration

### ISSUE THREAD

**Issue by ChuckDaniels87** [polygamma/aurman#16]
I've seen that you have already some tests. It would be nice to have continuous integration into this project early on.

I did some preliminary work in [this branch](https://github.com/ChuckDaniels87/pacaur/tree/ci-tests) for `pacaur` project. There is a custom Docker image that could be adapted and Travis CI simple config. If you are interested I could give a hand with this.

**Comment by polygamma** [comment]
I haven't worked with docker or continuous integration in general. But looking at both of those has been on my to-do-list for quite a long time, so I guess now I've got a good reason to do that. So that will happen, but no ETA on that...

**Comment by craftyguy** [comment]
There's also [arch-travis](https://github.com/mikkeloscar/arch-travis), I use this in one of my projects since travis-ci doesn't provide an Arch option, and I found that there were too many differences between the crappy Ubuntu image in travis-ci and Arch to make it reliable.

I can also give a hand when you want to implement this!

**Comment by ChuckDaniels87** [comment]
Well, you can use docker in the Travis Ubuntu image. I just simply generate a clean arch image from arch base and docker-run needed tests inside it.

**Comment by polygamma** [comment]
I think I want to start with this over the next couple of days. Since I've never worked with docker in general, any recommendations for documentation about docker? Besides the official docker documentation and the arch wiki of course ;)

**Comment by ChuckDaniels87** [comment]
It is quite easy to use, you just need an image from DockerHub to start running tests in a controlled environment. You can build that image or take an existing one.

**Comment by polygamma** [comment]
see: https://github.com/polygamma/aurman/commit/d9dfbd188c2d683aa47c381a4402515ec543a9fc

**Comment by ChuckDaniels87** [comment]
You can also use matrix/env variables to run several independent tests, and also to organize them in related batches. 
[More info](https://docs.travis-ci.com/user/environment-variables/)
[Simple example](https://github.com/ChuckDaniels87/pacaur/blob/ci-tests/.travis.yml)

PS: Not sure why you put a `sh` script with COPY and RUN it in the DockerFile.

**Comment by polygamma** [comment]
I just like the code more that way. Not to fiddle with things like `docker run -v $(pwd):/home/pacaur/pacaur arch-pacaur /bin/bash -lc "sudo chown -R pacaur .; $TEST"`

**Comment by ChuckDaniels87** [comment]
If I recall correctly I had to do some ugly hacks since I needed to build cower first to test `pacaur` install process. You don't need to use _ugly commands_ to use a matrix of tests:
```yaml
env:
    - TEST="script_1.py"
    - TEST="script_2.py"
script:
    - python $TEST
```

Some questions:
* How do you test install process if `aurman` is already installed in the docker image?
* If I understand correctly the current config, you are running some tests in Ubuntu image and then you will run other tests all together in `aurman` image running `docker_tests.sh`.

**Comment by polygamma** [comment]
First question: I already changed that, look at the latest commit.
Second question: That is not a question?

**Comment by ChuckDaniels87** [comment]
I can see in the last commit (1d68d6f4835cc9020813eefcb9249cb748c25717) that `aurman` is installed during Docker image generation and not in the script part.

To the second question: you can add -> did I get it right? :laughing:

**Comment by polygamma** [comment]
Ah, I get what you mean... I changed that, but I have a big problem now

see: https://github.com/polygamma/aurman/commit/09b9f80df74fe355e4391ebc74343b80bc07b900 and the test result: https://travis-ci.org/polygamma/aurman/builds/351756261

it has to do with the use of subprocess.run, that is how i execute the shell commands with aurman. but inside docker $PATH or whatever is not available, so that things like pod2man may not be found. i have no idea what to do...

and second question: yep

**Comment by polygamma** [comment]
Got it...

https://github.com/polygamma/aurman/compare/50c4831202951fb81b425350f78b2003e5532f01...2863ce9c63b5f53e72593f58e840e55952aed3c9

**Comment by polygamma** [comment]
Update: Works fine by now.
see: [Dockerfile](https://github.com/polygamma/aurman/blob/master/Dockerfile) and [docker_tests.sh](https://github.com/polygamma/aurman/blob/master/src/docker_tests/docker_tests.sh)

When creating the docker container, I am just pulling the files of the branch into that container and setting some permissions. The entrypoint is a for loop executing all .sh files inside the docker_tests folder. So currently one test is being executed.

In that test, I am installing aurman, cower, then cower-git, which removes cower, use aurmansolver for ros-indigo-desktop-full and mingw-w64-gcc, install fprintd and libfprint-vfs0090-git, which may require bootstrapping depending on the order of installing those, and finally installing pacman-git which removes pacman

**Comment by ChuckDaniels87** [comment]
Looks much better. :+1:

**Comment by polygamma** [comment]
Thanks :)

Do you have ideas for sensible tests all in all, @ChuckDaniels87 ?

**Comment by ChuckDaniels87** [comment]
For instance, cleaning cache folder, ignore packages, etc. Just standard operations and config options that should work in any case.

**Comment by polygamma** [comment]
I guess I'll switch to python tests, doing anything of that with bash seems like 3000% pain.

**Comment by ChuckDaniels87** [comment]
It seems that test system in 0abbfd6a5e010c5c28fca62127ced966f0350e0d is a bit convoluted. If you want to use Python instead of Bash for some reason, make system calls directly from the Python script.

**Comment by polygamma** [comment]
I know, I am just trying things right now, since that docker thing is new for me. I am going to rewrite everything of that

**Comment by polygamma** [comment]
@ChuckDaniels87 - https://github.com/polygamma/aurman/commit/6423d6f09adf5bc9f6ee11dc4c81f5b8c93d1a98

**Comment by ChuckDaniels87** [comment]
@polygamma much better. In fact, now you can easily use Travis test matrix to run tests independently and in parallel, as I told you [here](https://github.com/polygamma/aurman/issues/16#issuecomment-372039865).
```yaml
env:
    - TEST="src/unit_tests/test_split_query_helper.py"
    - TEST="src/unit_tests/test_parse_pacman_args.py"
    - TEST="src/docker_tests/install_tests.py"
    - TEST="src/docker_tests/cache_tests.py"
script:
    - docker run aurman-docker $TEST
```
This will run all tests in parallel, and they could fail independently without messing with each other. This solution would be potentially faster if you run long tests.

DISCLOSURE: I moved all script execution to the Arch image but I don't know if that will work. I think it is better to test ALL in an Arch environment. I did not test this config, it may kill your cat.

Great work! :+1:

**Comment by polygamma** [comment]
Going to try that out, thanks :)

**Comment by polygamma** [comment]
Well, it's actually not that easy.

- splitting tests into parallel tasks changes how the built docker images may be used, just doing what you suggested leads to not being able to use the image without the use of "docker login"
- just executing the normal unit tests may not be done with that syntax, because of how i splitted the aurman into packages, one needs to execute with python -m aurman. ...

I think currently it is perfectly fine to do it with the way it is, it's not like one really needs to execute those things parallel. also doing the normal unit tests outside a docker seems fine, too. using docker for that is overkill, really.

**Comment by ChuckDaniels87** [comment]
> I think currently it is perfectly fine to do it the way it is, it's not like one really needs to execute those things parallel. also doing the normal unit tests outside a docker container seems fine, too. using docker for that is overkill, really.

Fair enough! Well, continuous integration seems to be fully working now. I suppose you can close this issue.

**Comment by polygamma** [comment]
:) Your initial code for pacaur and your responses in this issue have been really helpful! Thank you very much!

**Comment by polygamma** [comment]
Added tests for custom build directory: https://github.com/polygamma/aurman/compare/2.9.2...cfe9d719f9221018a5924e224a554d92bd290b23

**Comment by polygamma** [comment]
Nevertheless status update, may be interesting for @ChuckDaniels87 ,too...
Being checked:

- installation of packages
- reset arch to 30 days ago and update with aurman `--do_everything`
- check building with custom build directory with pacman and pacman-git
- check cache cleaning
- check adding other repos, `testing` in that case, and update with aurman `--do_everything`
